src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -42,8 +42,15 @@ private static IEnumerable<TSource> WhereIterator<TSource>(IEnumerable<TSource>
             if (source == null) throw Error.ArgumentNull("source");
             if (selector == null) throw Error.ArgumentNull("selector");
             if (source is Iterator<TSource>) return ((Iterator<TSource>)source).Select(selector);
-            if (source is TSource[]) return new WhereSelectArrayIterator<TSource, TResult>((TSource[])source, null, selector);
-            if (source is List<TSource>) return new WhereSelectListIterator<TSource, TResult>((List<TSource>)source, null, selector);
+            IList<TSource> ilist = source as IList<TSource>;
+            if (ilist != null)
```

Comment:
This reduces the cost of testing for `TSource[]` or `List<TSource>` in the case where the source is not such a type, by one check, increases by one check when they are of such a type, but also adds support for other `IList<TSource>` types.